File: 496.next-greater-element-i.cpp
class Solution
{
public:
    vector<int> nextGreaterElement(vector<int> &nums1, vector<int> &nums2)
    {
        vector<int> res;
        res.reserve(nums1.size());
        stack<int> s;                  // monotonic decreasing stack
        unordered_map<int, int> table; //record the next greater element for all elements in nums2

        for (const int num : nums2)
        {
            if (s.empty())
                s.push(num);
            else if (num <= s.top())
            {
                s.push(num);
            }
            else
            {
                while (!s.empty() && s.top() < num)
                {
                    table[s.top()] = num;
                    s.pop();
                } //while
                s.push(num);
            }
        } // for

        // do all querys in nums
        for (const int x : nums1)
        {
            if (table.find(x) == table.end())
            {
                res.push_back(-1);
            } //if
            else
            {
                res.push_back(table[x]);
            } //else
        }     //for

        return res;
    }
};

File: Stack/496. Next Greater Element I/496. Next Greater Element I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2)
    {
        // use a stack to find NGE for all elts in nums2
        Stack<Integer> stack = new Stack<>();
        Map<Integer, Integer> mp = new HashMap<>();
        for(int i = nums2.length-1; i >= 0; --i)
        {
            while(!stack.isEmpty() && nums2[i]>= stack.peek())
                stack.pop();

            if(stack.empty())
                mp.put(nums2[i], -1);
            else
                mp.put(nums2[i], stack.peek());
            stack.push(nums2[i]);
        }

        for(int i = 0; i < nums1.length; ++i)
            nums1[i] = mp.get(nums1[i]);

        return nums1;
    }
}

// find Next Greater => non-increasing
// find next smaller => non-decreasing
// nums2: [1,3,-1, -1]
// stack: 4, 3
//

